File: hydra-java/src/main/java/hydra/dsl/LiteralTypes.java
package hydra.dsl;

import hydra.core.FloatType;
import hydra.core.IntegerType;
import hydra.core.LiteralType;


/**
 * DSL utilities for working with literal types.
 */
public interface LiteralTypes {
    static LiteralType bigfloat() {
        return float_(new FloatType.Bigfloat(false));
    }

    static LiteralType bigint() {
        return integer(new IntegerType.Bigint(false));
    }

    static LiteralType binary() {
        return new LiteralType.Binary(false);
    }

    static LiteralType boolean_() {
        return new LiteralType.Boolean_(false);
    }

    static LiteralType float32() {
        return float_(new FloatType.Float32(false));
    }

    static LiteralType float64() {
        return float_(new FloatType.Float64(false));
    }

    static LiteralType float_(final FloatType ftype) {
        return new LiteralType.Float_(ftype);
    }

    static LiteralType int16() {
        return integer(new IntegerType.Int16(false));
    }

    static LiteralType int32() {
        return integer(new IntegerType.Int32(false));
    }

    static LiteralType int64() {
        return integer(new IntegerType.Int64(false));
    }

    static LiteralType int8() {
        return integer(new IntegerType.Int8(false));
    }

    static LiteralType integer(final IntegerType itype) {
        return new LiteralType.Integer_(itype);
    }

    static LiteralType string() {
        return new LiteralType.String_(false);
    }

    static LiteralType uint16() {
        return integer(new IntegerType.Uint16(false));
    }

    static LiteralType uint32() {
        return integer(new IntegerType.Uint32(false));
    }

    static LiteralType uint64() {
        return integer(new IntegerType.Uint64(false));
    }

    static LiteralType uint8() {
        return integer(new IntegerType.Uint8(false));
    }

    /**
     * Encode a literal type as a string.
     */
    static String showLiteralType(LiteralType type) {
        return type.accept(new LiteralType.Visitor<String>() {
            @Override
            public String visit(LiteralType.Binary instance) {
                return "binary";
            }

            @Override
            public String visit(LiteralType.Boolean_ instance) {
                return "boolean";
            }

            @Override
            public String visit(LiteralType.Float_ instance) {
                return "float:" + instance.value.accept(new FloatType.Visitor<String>() {
                    @Override
                    public String visit(FloatType.Bigfloat instance) {
                        return "bigfloat";
                    }

                    @Override
                    public String visit(FloatType.Float32 instance) {
                        return "float32";
                    }

                    @Override
                    public String visit(FloatType.Float64 instance) {
                        return "float64";
                    }
                });
            }

            @Override
            public String visit(LiteralType.Integer_ instance) {
                return "integer:" + instance.value.accept(new IntegerType.Visitor<String>() {
                    @Override
                    public String visit(IntegerType.Bigint instance) {
                        return "bigint";
                    }

                    @Override
                    public String visit(IntegerType.Int8 instance) {
                        return "int8";
                    }

                    @Override
                    public String visit(IntegerType.Int16 instance) {
                        return "int16";
                    }

                    @Override
                    public String visit(IntegerType.Int32 instance) {
                        return "int32";
                    }

                    @Override
                    public String visit(IntegerType.Int64 instance) {
                        return "int64";
                    }

                    @Override
                    public String visit(IntegerType.Uint8 instance) {
                        return "uint8";
                    }

                    @Override
                    public String visit(IntegerType.Uint16 instance) {
                        return "uint16";
                    }

                    @Override
                    public String visit(IntegerType.Uint32 instance) {
                        return "uint32";
                    }

                    @Override
                    public String visit(IntegerType.Uint64 instance) {
                        return "uint64";
                    }
                });
            }

            @Override
            public String visit(LiteralType.String_ instance) {
                return "string";
            }
        });
    }
}

File: hydra-haskell/src/main/haskell/Hydra/Dsl/LiteralTypes.hs
-- | A DSL for Hydra literal types in Haskell

module Hydra.Dsl.LiteralTypes where

import Hydra.Core


-- | Arbitrary-precision floating point type
-- Example: bigfloat
bigfloat :: LiteralType
bigfloat = float FloatTypeBigfloat

-- | Arbitrary-precision integer type
-- Example: bigint
bigint :: LiteralType
bigint = LiteralTypeInteger IntegerTypeBigint

-- | Binary data type
-- Example: binary
binary :: LiteralType
binary = LiteralTypeBinary

-- | Boolean type
-- Example: boolean
boolean :: LiteralType
boolean = LiteralTypeBoolean

-- | 32-bit floating point type
-- Example: float32
float32 :: LiteralType
float32 = float FloatTypeFloat32

-- | 64-bit floating point type
-- Example: float64
float64 :: LiteralType
float64 = float FloatTypeFloat64

-- | Create a floating point type with the specified precision
-- Example: float FloatTypeFloat32
float :: FloatType -> LiteralType
float = LiteralTypeFloat

-- | 16-bit signed integer type
-- Example: int16
int16 :: LiteralType
int16 = integer IntegerTypeInt16

-- | 32-bit signed integer type
-- Example: int32
int32 :: LiteralType
int32 = integer IntegerTypeInt32

-- | 64-bit signed integer type
-- Example: int64
int64 :: LiteralType
int64 = integer IntegerTypeInt64

-- | 8-bit signed integer type
-- Example: int8
int8 :: LiteralType
int8 = integer IntegerTypeInt8

-- | Create an integer type with the specified bit width
-- Example: integer IntegerTypeInt32
integer :: IntegerType -> LiteralType
integer = LiteralTypeInteger

-- | String type
-- Example: string
string :: LiteralType
string = LiteralTypeString

-- | 16-bit unsigned integer type
-- Example: uint16
uint16 :: LiteralType
uint16 = integer IntegerTypeUint16

-- | 32-bit unsigned integer type
-- Example: uint32
uint32 :: LiteralType
uint32 = integer IntegerTypeUint32

-- | 64-bit unsigned integer type
-- Example: uint64
uint64 :: LiteralType
uint64 = integer IntegerTypeUint64

-- | 8-bit unsigned integer type
-- Example: uint8
uint8 :: LiteralType
uint8 = integer IntegerTypeUint8
